### PR TITLE
NEW FEATURE: improve/fix scipy.minimize interfaces in wec.solve

### LINKED ISSUE #146: core: allow `callback` to use the same function signature as the other functions.
Use the standard `fun(wec, x_wec, x_opt, waves)` signature for callbacks instead of the `fun(x)` currently used.

### LINKED ISSUE #148: Core: allow for all options of `bounds`
See `scipy.optimize.minimize` documentation: https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html

### LINKED ISSUE #157: core: unscale all the fields of `res`
We currently unscale `x` and `fun` from the [`OptimizeResult`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.OptimizeResult.html#scipy.optimize.OptimizeResult), but not `jac`, `hess`, etc.

https://github.com/SNL-WaterPower/WecOptTool/blob/09f6ede87628f8c5a735fff0c368497afc8e171a/wecopttool/core.py#L755-L758

Related to #4

### PR DESCRIPTION
**Pull request by ryancoe**
## Description
Could be called a bug fix because of #157, but probably more appropriate to call this a new feature change.

- [x] Closes #146 
- [x] Closes #148
- [x] Closes #157 

## Checklist for PR
- [x] Authors read the [contribution guidelines](https://github.com/SNL-WaterPower/WecOptTool/blob/main/.github/CONTRIBUTING.md)
- [x] The pull request is from an issue branch (not main) on your fork to the [main branch in WecOptTool](https://github.com/SNL-WaterPower/WecOptTool).
- [x] The authors have given the admins edit access
- [x] All changes adhere to the style guide including PEP8, Docstrings, and Type Hints.
~~Modified the documentation if applicable~~
- [x] Modified or added a new test
- [x] All tests pass
- [x] Documentation builds
- [x] [Reference or close any relevant issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)

### CODE REVIEW

File: wecopttool/core.py [hunk 1]
```diff
@@ -755,7 +763,7 @@ def callback(x):
         # unscale
         optim_res.x = optim_res.x / scale
         optim_res.fun = optim_res.fun / scale_obj
-        # TODO: unscale all the other fields in the results, e.g. 'jac'
+        optim_res.jac = optim_res.jac / scale_obj * scale
```
Comment (ryancoe) [inline wecopttool/core.py:766]: see check logic in #157

Comment (cmichelenstrofer) [reply wecopttool/core.py:766]: Makes sense!

File: wecopttool/core.py [hunk 2]
```diff
@@ -721,12 +725,16 @@ def resid_fun(x):
...
+            def callback_scipy(x):
                 x_wec, x_opt = self.decompose_state(x)
                 _log.info("[max(x_wec), max(x_opt), obj_fun(x)]: "
                           + f"[{np.max(np.abs(x_wec)):.2e}, "
                           + f"{np.max(np.abs(x_opt)):.2e}, "
                           + f"{np.max(obj_fun_scaled(x)):.2e}]")
+        else:
+            def callback_scipy(x):
+                x_wec, x_opt = self.decompose_state(x)
+                return callback(self, x_wec, x_opt, waves)
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:737]: 👍🏾

File: wecopttool/core.py [hunk 3]
```diff
@@ -703,8 +704,11 @@ def resid_fun(x):
         if (bounds_wec is None) and (bounds_opt is None):
             bounds = None
         else:
-            # TODO: allow for all options of Bounds.
             bounds_in = [bounds_wec, bounds_opt]
+            for idx, bii in enumerate(bounds_in):
+                if isinstance(bii, tuple):
+                    bounds_in[idx] = Bounds(lb=[xibs[0] for xibs in bii], 
+                                            ub=[xibs[1] for xibs in bii])
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:711]: Nice!

File: tests/test_integration.py [hunk 4]
```diff
@@ -94,6 +104,26 @@ def wec_from_impedance(bem, pto, fb):
...
+    _ = wec_from_bem.solve(regular_wave,
+                  obj_fun=pto.average_power,
+                  nstate_opt=2*nfreq+1,
+                  scale_x_wec=1.0,
+                  scale_x_opt=0.01,
+                  scale_obj=1e-1,
+                  callback=my_callback,
+                  optim_options={'maxiter': 1})
+
+    out, err = capfd.readouterr()
```
Comment (cmichelenstrofer) [inline tests/test_integration.py:124]: What is capfd? Don't see it defined anywhere. Does this run?

**Review by cmichelenstrofer — APPROVED** [review]

File: tests/test_integration.py (hunk 4 above)
Comment (ryancoe) [reply tests/test_integration.py:124]: Yes, not something I'd done before but here's how it works: https://docs.pytest.org/en/7.1.x/how-to/capture-stdout-stderr.html#accessing-captured-output-from-a-test-function

Comment (cmichelenstrofer) [reply tests/test_integration.py:124]: Oh cool! So they are fixtures already defined by default.

**Pull request merged**